Add support for encoding/decoding feature ids

victorzinho opened the pull request:
[Feature ids](https://github.com/mapbox/vector-tile-spec/blob/master/2.1/vector_tile.proto#L32) were not supported. I added the possibility to specify the feature id either in the `addFeature` method or with an `autoincrement` flag in the encoder constructor and then write those ids in the `encode` method.

I added it as new methods/constructors with disabled functionality by default to ensure backwards compatibility.

jratike80 commented:
Vector tile specification seems to require that id is of type uint64. Features which are delivered by Geoserver typically require also the name of the feature type for making them universally unique, like
gml:id="states.1"

If ids for MVT are created somehow on-the-fly, how can we guarantee that a big polygon that is saved partially into one MVT tile will get the same id when another part of it is saved into another MVT tile sometimes later?

victorzinho commented:
The autoincrement feature I added was just in case you need IDs to identify features in the client but you don't care about those IDs because you won't modify anything. Not my specific use case, just added sugar.

For GeoServer I just extract the numeric portion of the GeoServer fid if possible (like [this](https://github.com/geomatico/geoserver/commit/4cfabca322284c548f5c169c1b89999b5a52f0dd)) so the FIDs are consistent. I compiled the library (java-vector-tile) and extension (gs-vectortiles) and tested locally the changes and it's working properly with Mapbox GL JS.

jratike80 commented:
There is discussion about if ids may be strings in MVT 3 version https://github.com/mapbox/vector-tile-spec/issues/94.
I guess that taking the numeric portion of Geoserver fid would work well when the features are coming from a single layer. I do not know if Geoserver can generate MVT tiles from a layer group but if it can then it may possible to get duplicate numeric portions from different native layers.

oscarfonts commented:
It would be nice to use the original Geoserver string identifier in MVT if/when they decide to support them.

Mapbox GL API asks for 3 parameters to identify a feature: its numeric feature.id, but also its source and sourceLayer. Just because the numeric id is not unique across layers (see https://www.mapbox.com/mapbox-gl-js/api#map#setfeaturestate).

Indeed, we're using this patch with 45-layer MVTs generated from a GeoServer layergroup. :)

victorzinho commented:
@halset This PR has been open for quite some time and now it has some conflicts that require some work.

It would be great if I could get some feedback before resolving the conflicts, just to know whether I need to keep working on it or just drop it. 

Thanks in advance.

halset commented:
Sorry for this. I have merged you work in to master now. Thank you very much! I will have to create a release and also maintain the jts-vividsolutions branch.

halset commented:
Backported to the jts-vividsolutions branch and made releases. 1.3.3 (master) and 1.2.3 (jts-vividsolutions).
